pkg/vsphere/vm/vm.go

```diff
@@ -427,3 +425,110 @@ func (vm *VirtualMachine) UpgradeInProgress(ctx context.Context, upgradePrefix s
 
 	return false, "", nil
 }
+
+func (vm *VirtualMachine) registerVM(ctx context.Context, path, name string,
+	vapp, pool, host *types.ManagedObjectReference, vmfolder *object.Folder) (*object.Task, error) {
+	log.Debugf("Register VM %s", name)
+
+	if vapp == nil {
+		var hostObject *object.HostSystem
+		if host != nil {
+			hostObject = object.NewHostSystem(vm.Vim25(), *host)
+		}
+		poolObject := object.NewResourcePool(vm.Vim25(), *pool)
+		return vmfolder.RegisterVM(ctx, path, name, false, poolObject, hostObject)
+	}
+
+	req := types.RegisterChildVM_Task{
+		This: vapp.Reference(),
+		Path: path,
+		Host: host,
+	}
+
+	if name != "" {
+		req.Name = name
+	}
+
+	res, err := methods.RegisterChildVM_Task(ctx, vm.Vim25(), &req)
+	if err != nil {
+		return nil, err
+	}
+
+	return object.NewTask(vm.Vim25(), res.Returnval), nil
+}
+
+// FixInvalidState fix vm invalid state issue through unregister & register
+func (vm *VirtualMachine) FixInvalidState(ctx context.Context) error {
+	log.Debugf("Fix invalid state VM: %s", vm.Reference())
+	folders, err := vm.Session.Datacenter.Folders(ctx)
+	if err != nil {
+		log.Errorf("Unable to get vm folder: %s", err)
+		return err
+	}
+
+	properties := []string{"config.files", "summary.config", "summary.runtime", "resourcePool", "parentVApp"}
+	log.Debugf("Get vm properties %s", properties)
+	var mvm mo.VirtualMachine
+	if err = vm.Properties(ctx, vm.Reference(), properties, &mvm); err != nil {
+		log.Errorf("Unable to get vm properties: %s", err)
+		return err
+	}
+
+	name := mvm.Summary.Config.Name
+	log.Debugf("Unregister VM %s", name)
+	if err := vm.Unregister(ctx); err != nil {
+		log.Errorf("Unable to unregister vm %q: %s", name, err)
+		return err
+	}
+
+	task, err := vm.registerVM(ctx, mvm.Config.Files.VmPathName, name, mvm.ParentVApp, mvm.ResourcePool, mvm.Summary.Runtime.Host, folders.VmFolder)
+	if err != nil {
+		log.Errorf("Unable to register VM %q back: %s", name, err)
+		return err
+	}
+	info, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		return err
+	}
+	// re-register vm will change vm reference, so reset the object reference here
+	if info.Error != nil {
+		return errors.New(info.Error.LocalizedMessage)
```

Comment:
In this case, err is nil, but info.Error is not. I doubt this will happen, but if it does, this error might be better.